File: src/main/java/workshop/example/TokenValidator.java
package example;

import com.nimbusds.jose.jwk.source.RemoteJWKSet;
import com.nimbusds.jose.proc.*;
import com.nimbusds.jwt.proc.*;
import java.net.URL;
import java.util.*;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.jwk.source.JWKSource;
import com.nimbusds.jose.jwk.source.JWKSourceBuilder;

public class TokenValidator {

	private final ConfigurableJWTProcessor<SecurityContext> processor;

	public TokenValidator(String jwksUrl) throws Exception {
		JWKSource<SecurityContext> keySource = new RemoteJWKSet<>(new URL(jwksUrl));

		processor = new DefaultJWTProcessor<>();
		processor.setJWSKeySelector(new JWSVerificationKeySelector<>(JWSAlgorithm.ES256, keySource));

		JWTClaimsSet requiredClaims = new JWTClaimsSet.Builder().issuer("https://auth.fintech.com").build();

		processor.setJWTClaimsSetVerifier(new DefaultJWTClaimsVerifier<>(requiredClaims,
				new HashSet<>(Arrays.asList("sub", "exp", "iat", "jti", "scope"))));
	}

	public JWTClaimsSet validate(String token) throws Exception {
		return processor.process(token, null);
	}

	public List<String> extractScopes(JWTClaimsSet claims) throws Exception {
		String scope = claims.getStringClaim("scope");
		return scope != null ? Arrays.asList(scope.split(" ")) : Collections.emptyList();
	}
}

File: src/main/scala/workshop/TokenValidator.scala
package workshop

import java.net.URI

import scala.jdk.CollectionConverters.*
import scala.util.Try

import com.nimbusds.jose.jwk.source.{JWKSource, JWKSourceBuilder}
import com.nimbusds.jose.proc.{JWSVerificationKeySelector, SecurityContext}
import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jwt.proc.{
  ConfigurableJWTProcessor,
  DefaultJWTClaimsVerifier,
  DefaultJWTProcessor
}
import com.nimbusds.jwt.JWTClaimsSet

class TokenValidator(jwksUrl: String) {

  private val keySource: JWKSource[SecurityContext] =
    JWKSourceBuilder
      .create[SecurityContext](URI.create(jwksUrl).toURL())
      .build()

  private val processor: ConfigurableJWTProcessor[SecurityContext] = {
    val p = new DefaultJWTProcessor[SecurityContext]()
    p.setJWSKeySelector(
      new JWSVerificationKeySelector(JWSAlgorithm.ES256, keySource)
    )

    // Require specific claims
    val requiredClaims = new JWTClaimsSet.Builder()
      .issuer("https://auth.fintech.com")
      .build()
    p.setJWTClaimsSetVerifier(
      new DefaultJWTClaimsVerifier(requiredClaims, Set("sub", "exp").asJava)
    )
    p
  }

  // 2. Validation method
  def validate(token: String): Either[Throwable, JWTClaimsSet] =
    Try(processor.process(token, null)).toEither

}

// 3. Idiomatic Scala 3 Extension Methods for safe claim extraction
extension (claims: JWTClaimsSet) {

  def getScopeList: List[String] =
    Option(claims.getStringClaim("scope"))
      .map(_.split(" ").toList)
      .getOrElse(List.empty)

  def getOptionalClaim(name: String): Option[String] =
    Option(claims.getStringClaim(name))

}

// Usage:
// val result = validator.validate(tokenStr)
// result.map(_.getScopeList).foreach(println)
